### PR TITLE
update pdfcpu to version 0.4.0 (+ fixes)

### PR DESCRIPTION
**Pull request by christian-heusel**
A new version of the great pdfcpu has just been released: https://github.com/pdfcpu/pdfcpu/releases/tag/v0.4.0

---

**To be fixed before merge/release:** 
 - [x] fix usage of `api.Merge()` (I don't think I can do this):
   https://github.com/henrixapp/pdfcomprezzor/blob/8d4dce38599eb3457dc582b484da3996f1b0e2c1/main.go#L107
   ```
   GOOS=js GOARCH=wasm go build -o pdfcomprezzor.wasm                                                                                                1 ↵

   # github.com/henrixapp/pdfcomprezzor
   ./main.go:108:25: not enough arguments in call to api.Merge
           have ([]io.ReadSeeker, *bytes.Buffer, nil)
           want (string, []string, io.Writer, *model.Configuration)
   ```
- [x] wait for PR https://github.com/pdfcpu/pdfcpu/pull/565 to be merged

### CODE REVIEW

**Comment by henrixapp** [comment]
> fix usage of api.Merge() (I don't think I can do this):

The API was changed in https://github.com/pdfcpu/pdfcpu/commit/a86469b721d0592fd1b5a21bb46aa969ef9d7e74 , we will have to wait for a update on pdfcpu site.

**Comment by christian-heusel** [comment]
> The API was changed in https://github.com/pdfcpu/pdfcpu/commit/a86469b721d0592fd1b5a21bb46aa969ef9d7e74 

I also noticed that and used `api.MergeOld()` but that got removed in the meantime aswell.

> we will have to wait for a update on pdfcpu site.

Can you report an issue about it and link it here? I unfortunately cant argue why things are the way they are or what exactly needs to be changed in order to get it running again 🤔

**Comment by christian-heusel** [comment]
@henrixapp I think this is ready for merge, please review the changes, test and merge 👍🏻

**Comment by christian-heusel** [comment]
Regarding the replace: We could ofc also wait for an upstream relase, but the last one took a year xD

**Comment by christian-heusel** [comment]
Now I sadly have a found regression, with the pdf I sent you via messenger I get the following error on this branch: 
```
GEThttp://localhost:8082/favicon.ico
[HTTP/1.1 404 Not Found 0ms]

1 test.js:32:25
Worker said:  
message { target: Worker, isTrusted: true, data: {…}, origin: "", lastEventId: "", ports: Restricted, srcElement: Worker, currentTarget: Worker, eventPhase: 2, bubbles: false, … }
test.js:40:33
Worker said:  
message { target: Worker, isTrusted: true, data: {…}, origin: "", lastEventId: "", ports: Restricted, srcElement: Worker, currentTarget: Worker, eventPhase: 2, bubbles: false, … }
test.js:40:33
Worker said:  
message { target: Worker, isTrusted: true, data: {…}, origin: "", lastEventId: "", ports: Restricted, srcElement: Worker, currentTarget: Worker, eventPhase: 2, bubbles: false, … }
test.js:40:33
Worker said:  
message { target: Worker, isTrusted: true, data: {…}, origin: "", lastEventId: "", ports: Restricted, srcElement: Worker, currentTarget: Worker, eventPhase: 2, bubbles: false, … }
test.js:40:33
Worker said:  
message { target: Worker, isTrusted: true, data: {…}, origin: "", lastEventId: "", ports: Restricted, srcElement: Worker, currentTarget: Worker, eventPhase: 2, bubbles: false, … }
test.js:40:33
Worker said:  
message { target: Worker, isTrusted: true, data: {…}, origin: "", lastEventId: "", ports: Restricted, srcElement: Worker, currentTarget: Worker, eventPhase: 2, bubbles: false, … }
test.js:40:33
Worker said:  
message { target: Worker, isTrusted: true, data: {…}, origin: "", lastEventId: "", ports: Restricted, srcElement: Worker, currentTarget: Worker, eventPhase: 2, bubbles: false, … }
test.js:40:33
Worker said:  
message { target: Worker, isTrusted: true, data: {…}, origin: "", lastEventId: "", ports: Restricted, srcElement: Worker, currentTarget: Worker, eventPhase: 2, bubbles: false, … }
test.js:40:33
Worker said:  
message { target: Worker, isTrusted: true, data: {…}, origin: "", lastEventId: "", ports: Restricted, srcElement: Worker, currentTarget: Worker, eventPhase: 2, bubbles: false, … }
test.js:40:33
panic: runtime error: index out of range [9] with length 1 wasm_exec.js:22:14
<empty string> wasm_exec.js:22:14
goroutine 6 [running]: wasm_exec.js:22:14
main.glob..func1({{}, 0x0, 0x0}, {0x82c3e0, 0x2, 0x2}) wasm_exec.js:22:14
	/go/src/main.go:63 +0x126 wasm_exec.js:22:14
syscall/js.handleEvent() wasm_exec.js:22:14
	/usr/local/go/src/syscall/js/func.go:94 +0x26 wasm_exec.js:22:14
exit code: 2 wasm_exec.js:101:14
Error: Go program has already exited wasm_exec.js:536:11
```

File: main.go [hunk 1]
```diff
@@ -46,13 +47,13 @@ var onCompress = js.FuncOf(func(this js.Value, args []js.Value) interface{} {
 
 	for i := 1; i <= count; i++ {
 		Log("Processing page no:", i)
-		imageObjNrs := ctx.ImageObjNrs(i)
+		imageObjNrs := pdfcpu.ImageObjNrs(ctx, i)
 		Log("Images on page:", len(imageObjNrs))
-		images, err := ctx.ExtractPageImages(i, false)
+		images, err := pdfcpu.ExtractPageImages(ctx, i, false)
```
Comment (henrixapp) [inline main.go:52]: ExtractPageImages returns a map instead of an array. Please update the lines below.

File: main.go [hunk 2]
```diff
@@ -73,7 +74,7 @@ var onCompress = js.FuncOf(func(this js.Value, args []js.Value) interface{} {
 				}
 				buf := new(bytes.Buffer)
 				png.Encode(buf, smaller)
-				sd2, _, _, _ := pdfcpu.CreateImageStreamDict(ctx.XRefTable, buf, false, false)
+				sd2, _, _, _ := model.CreateImageStreamDict(ctx.XRefTable, buf, false, false)
 				ctx.XRefTable.Table[objs[idx]].Object = *sd2
```
Comment (henrixapp) [inline main.go]: Now it is just `idx` instead of `objs[idx]`.

Comment (henrixapp) [reply main.go]: also update above.

**Review by henrixapp — CHANGES_REQUESTED** [review]
Please update to reflect map return type instead of array.

**Review by henrixapp — APPROVED** [review]

**Pull request merged**